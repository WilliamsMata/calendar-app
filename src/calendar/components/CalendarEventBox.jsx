export const CalendarEventBox = ({ event }) => {
  const { title, user } = event;

  return (
    <>
      <strong>{title}</strong>
      <span> - {user.name}</span>
    </>
  );
};
