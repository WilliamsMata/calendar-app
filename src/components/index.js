export * from "./CheckingAuthSpinner";
export * from "./DarkModeBtn";
