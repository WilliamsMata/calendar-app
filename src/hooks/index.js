export * from "./useDarkMode";
export * from "./useLanguage";
export * from "./useUiStore";
