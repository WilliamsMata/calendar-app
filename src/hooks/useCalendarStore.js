import { useDispatch, useSelector } from "react-redux";
import {
  onAddNewEvent,
  onDeleteEvent,
  onSetActiveEvent,
  onUpdateEvent,
  onClearActiveEvent,
} from "../store";

export const useCalendarStore = () => {
  const dispatch = useDispatch();
  const { events, activeEvent } = useSelector((state) => state.calendar);

  const setActiveEvent = (calendarEvent) => {
    dispatch(onSetActiveEvent(calendarEvent));
  };

  const startSavingEvent = async (calendarEvent) => {
    //Todo: Llegar al backend

    // All good
    if (calendarEvent._id) {
      // Updating
      dispatch(onUpdateEvent({ ...calendarEvent }));
    } else {
      // Creating
      dispatch(onAddNewEvent({ ...calendarEvent, _id: new Date().getTime() }));
    }
  };

  const startDeletingEvent = () => {
    //Todo: Llegar al backend

    dispatch(onDeleteEvent());
  };

  const clearActiveEvent = () => {
    dispatch(onClearActiveEvent());
  };

  return {
    //* Properties
    activeEvent,
    events,
    hasEventSelected: !!activeEvent,

    //* Methods
    setActiveEvent,
    startSavingEvent,
    startDeletingEvent,
    clearActiveEvent,
  };
};
