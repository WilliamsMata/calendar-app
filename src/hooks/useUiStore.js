import { useDispatch, useSelector } from "react-redux";
import { onCloseDateModal, onOpenDateModal } from "../store";

export const useUiStore = () => {
  const { isDateModalOpen } = useSelector((state) => state.ui);
  const dispatch = useDispatch();

  const openDateModal = () => {
    dispatch(onOpenDateModal());
  };

  const closeDateModal = () => {
    dispatch(onCloseDateModal());
  };

  return {
    //* Properties
    isDateModalOpen,

    //* Methods
    openDateModal,
    closeDateModal,
  };
};
