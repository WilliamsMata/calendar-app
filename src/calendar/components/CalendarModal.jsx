import Modal from "react-modal";
import DatePicker, { registerLocale } from "react-datepicker";
import "react-datepicker/dist/react-datepicker.css";
import es from "date-fns/locale/es";
import { startOfDay } from "date-fns";
import "sweetalert2/dist/sweetalert2.min.css";

import {
  getEnvVariables,
  getFormMessageEN,
  getFormMessageES,
  isUserDeviceInSpanish,
} from "../../helpers";
import { useCalendarModal } from "../../hooks";
import { CloseModalBtn } from "./CloseModalBtn";
import { useSelector } from "react-redux";
import { LoaderSpinnerIcon } from "./LoaderSpinnerIcon";

registerLocale("es", es);

if (getEnvVariables().VITE_MODE !== "test") {
  Modal.setAppElement("#root");
}

const { titleForm, dateStart, dateEnd, eventTitle, note, colorLabel, button } =
  isUserDeviceInSpanish ? getFormMessageES() : getFormMessageEN();

export const CalendarModal = () => {
  const { isSavingEvent } = useSelector((state) => state.calendar);

  const {
    //* Properties
    formValues,
    isDateModalOpen,
    titleClass,
    activeEvent,
    colors,

    //* Methods
    onInputChange,
    onDateChange,
    onCloseModal,
    onSubmit,
    handleColorClick,
  } = useCalendarModal();

  return (
    <Modal
      isOpen={isDateModalOpen}
      onRequestClose={onCloseModal}
      className="relative mx-2 w-full max-w-[30rem] rounded-lg bg-base-100 p-4 outline-none"
      overlayClassName="modal-fondo"
      closeTimeoutMS={200}
    >
      <h1>{activeEvent?._id ? titleForm.editEvent : titleForm.newEvent}</h1>

      <form className="container form-control w-full" onSubmit={onSubmit}>
        <div className="border-b-2 border-base-200 pb-4">
          <div>
            <label className="label">{dateStart.label}</label>
            <DatePicker
              id="date-start"
              minDate={startOfDay(new Date())}
              selected={formValues.start}
              onChange={(event) => onDateChange(event, "start")}
              className="input-bordered input w-full"
              dateFormat="Pp"
              showTimeSelect
              locale={isUserDeviceInSpanish ? "es" : null}
              timeCaption={isUserDeviceInSpanish ? "Hora" : "Time"}
            />
          </div>

          <div>
            <label className="label">{dateEnd.label}</label>
            <DatePicker
              id="date-end"
              minDate={formValues.start}
              selected={formValues.end}
              onChange={(event) => onDateChange(event, "end")}
              className="input-bordered input w-full"
              dateFormat="Pp"
              showTimeSelect
              locale={isUserDeviceInSpanish ? "es" : null}
              timeCaption={isUserDeviceInSpanish ? "Hora" : "Time"}
            />
          </div>
        </div>

        <div className="flex flex-col gap-2">
          <div>
            <label htmlFor="event-title" className="label">
              {eventTitle.label}
            </label>
            <input
              id="event-title"
              type="text"
              className={`input-bordered input w-full ${titleClass}`}
              placeholder={eventTitle.placeholder}
              name="title"
              autoComplete="off"
              value={formValues.title}
              onChange={onInputChange}
            />
            <label
              htmlFor="event-title"
              id="emailHelp"
              className="label-text-alt"
            >
              {eventTitle.altLabel}
            </label>
          </div>

          <div>
            <textarea
              id="note-description"
              type="text"
              className="textarea-bordered textarea w-full resize-none"
              placeholder={note.placeholder}
              rows="3"
              name="notes"
              value={formValues.notes}
              onChange={onInputChange}
            />
            <label
              htmlFor="note-description"
              id="emailHelp"
              className="label-text-alt"
            >
              {note.altLabel}
            </label>
          </div>

          <div>
            <label className="label">{colorLabel}</label>
            <div className="mx-1 flex items-center justify-between gap-4">
              {colors.map((color) => (
                <div
                  key={color}
                  className={`h-8 w-16 cursor-pointer rounded-lg border ${
                    formValues.bgColor === color
                      ? `outline outline-2 outline-offset-2`
                      : ""
                  }`}
                  style={{ backgroundColor: color }}
                  onClick={() => handleColorClick(color)}
                />
              ))}
            </div>
          </div>

          <button
            id="save-btn"
            type="submit"
            className="btn-outline btn-info btn-block btn mt-2 disabled:bg-base-300"
            disabled={isSavingEvent}
          >
            {isSavingEvent ? (
              <LoaderSpinnerIcon />
            ) : (
              <>
                <img
                  src="./icons/save.svg"
                  alt="save logo"
                  className="mr-4 h-6 w-6 transition"
                />
                <span className="tracking-wider">{button}</span>
              </>
            )}
          </button>
        </div>
      </form>

      <CloseModalBtn />
    </Modal>
  );
};
