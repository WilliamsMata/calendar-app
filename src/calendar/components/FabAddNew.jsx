import { addDays, startOfDay } from "date-fns";
import { useCalendarStore, useUiStore } from "../../hooks";

export const FabAddNew = () => {
  const { openDateModal } = useUiStore();
  const { setActiveEvent } = useCalendarStore();

  const onAddNewClick = () => {
    setActiveEvent({
      title: "",
      notes: "",
      start: startOfDay(new Date()),
      end: startOfDay(addDays(new Date(), 1)),
      bgColor: "#661ae6",
      user: {
        _id: "123",
        name: "Williams",
      },
    });
    openDateModal();
  };

  return (
    <button
      id="add-new-btn"
      onClick={onAddNewClick}
      className="btn-info btn-circle btn-lg btn fixed right-4 bottom-4 transition hover:brightness-105 lg:right-8 lg:bottom-6"
    >
      <img
        src="./icons/add.svg"
        alt="add button"
        className="h-8 w-8 transition"
      />
    </button>
  );
};
