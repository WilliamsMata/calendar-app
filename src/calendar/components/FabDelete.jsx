import Swal from "sweetalert2";
import { isUserDeviceInSpanish } from "../../helpers";
import {
  getDeleteSweetModalMessageEN,
  getDeleteSweetModalMessageES,
} from "../../helpers/getMessages";
import { useCalendarStore, useUiStore } from "../../hooks";

const swalButtons = Swal.mixin({
  customClass: {
    confirmButton:
      "btn btn-info min-w-[5rem] mx-2 transition hover:brightness-110",
    cancelButton:
      "btn btn-error min-w-[5rem] mx-2 transition hover:brightness-110",
  },
  buttonsStyling: false,
});

const { question, cancelled } = isUserDeviceInSpanish
  ? getDeleteSweetModalMessageES()
  : getDeleteSweetModalMessageEN();

export const FabDelete = () => {
  const { isDateModalOpen } = useUiStore();
  const { startDeletingEvent, hasEventSelected, clearActiveEvent } =
    useCalendarStore();

  const onDelete = async () => {
    const result = await swalButtons.fire({
      title: question.title,
      text: question.text,
      icon: "warning",
      showCancelButton: true,
      confirmButtonText: question.confirmBtn,
      cancelButtonText: question.cancelBtn,
      reverseButtons: true,
    });

    if (result.isConfirmed) {
      await startDeletingEvent();
    } else if (result.dismiss === Swal.DismissReason.cancel) {
      swalButtons.fire(cancelled.title, cancelled.text, "error");
    }

    clearActiveEvent();
  };

  return (
    <button
      id="delete-btn"
      onClick={onDelete}
      className={`btn-error btn-circle btn-lg btn fixed left-4 bottom-4 transition hover:brightness-105 lg:left-8 lg:bottom-6 ${
        hasEventSelected ? "" : "hidden"
      } ${isDateModalOpen ? "hidden" : ""}`}
    >
      <img
        src="./icons/trash.svg"
        alt="delete button"
        className="h-8 w-8 transition"
      />
    </button>
  );
};
