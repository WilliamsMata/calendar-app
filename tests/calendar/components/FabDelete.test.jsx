import { render, screen } from "@testing-library/react";

import { FabDelete } from "../../../src/calendar/components/FabDelete";
import { useUiStore } from "../../../src/hooks/useUiStore";
import { useCalendarStore } from "../../../src/hooks/useCalendarStore";

jest.mock("../../../src/hooks/useUiStore");
jest.mock("../../../src/hooks/useCalendarStore");

describe("Test in <FabDelete />", () => {
  test("Should display the component correctly", () => {
    useUiStore.mockReturnValue({ isDateModalOpen: false });
    useCalendarStore.mockReturnValue({ hasEventSelected: false });

    const { container } = render(<FabDelete />);
    const btn = screen.getByRole("button");

    expect(container).toMatchSnapshot();
    expect(btn.classList).toContain("hidden");
  });

  test("should show the button if there is a event selected", () => {
    useUiStore.mockReturnValue({ isDateModalOpen: false });
    useCalendarStore.mockReturnValue({ hasEventSelected: true });

    render(<FabDelete />);
    const btn = screen.getByRole("button");

    expect(btn.classList).not.toContain("hidden");
  });
});
