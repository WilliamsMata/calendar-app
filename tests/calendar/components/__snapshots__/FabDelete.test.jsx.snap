// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in <FabDelete /> Should display the component correctly 1`] = `
<div>
  <button
    class="btn-error btn-circle btn-lg btn fixed left-4 bottom-4 transition hover:brightness-105 lg:left-8 lg:bottom-6 hidden "
    id="delete-btn"
  >
    <img
      alt="delete button"
      class="h-8 w-8 transition"
      src="./icons/trash.svg"
    />
  </button>
</div>
`;
